### ISSUE TITLE
UndocumentedPublicProperty doesn't recognize inline documentation in the primary constructor.

### ISSUE THREAD

**Issue by arhont375** [detekt/detekt#3677]
## Expected Behavior
When property have inline documentation in the primary constructor `UndocumentedPublicProperty` shouldn't fail.

## Observed Behavior

`UndocumentedPublicProperty` rule expect property defined in primary constructor to have documentation on Class level via `@param/@property`.

## Steps to Reproduce

1. Gradle plugins: 
```
    id("org.jetbrains.kotlin.jvm") version "1.4.32"
    id("io.gitlab.arturbosch.detekt") version "1.16.0"
```
2. Detekt configuration:
```groovy
// In gradle 
detekt {
    config = rootProject.files("config/detekt/detekt.yml")
    buildUponDefaultConfig = true // preconfigure defaults
}
```
```yaml
// In config file
comments:
  active: true
  UndocumentedPublicClass:
    active: true
  UndocumentedPublicProperty:
    active: true
```
3. Following class:
```kotlin
/**
 * This class have some KDoc.
 *
 * @property anotherImportantProperty Some documentation
 */
class TestProperties(
    /**
     * Some documentation
     */
    val seriousProperty: String,
    val anotherImportantProperty: Int,
) {
    /**
     * Some documentation
     */
    val justProperty: String = seriousProperty + seriousProperty
}
```
With described above settings Detekt will report `UndocumentedPublicProperty` for the `seriousProperty` even though documentation provided. Presence or absence of class level documentation doesn't change behavior.

## Context
Class may be relatively big or may have long description on the class level about details of implementation, in that case for someone reading the class it's much easier to see docs if they just next to the property declaration.

## Your Environment

* Version of detekt used: 1.16.0
* Version of Gradle used (if applicable): 6.8.3
* Gradle scan link (add `--scan` option when running the gradle task): <omitted>
* Operating System and version: Mac OS 10.15.7
* Link to your project (if it's a public repository): <omitted>

**Comment by schalkms** [comment]
The semantics of the mentioned rule complies with the official KDoc format. What you describe does not comply with the linked guide [1].
Hence, your given code snippet is flagged, which is correct system behavior.
As documented in [1], the primary constructor documentation is introduced with the `@constructor` tag.

[1] https://kotlinlang.org/docs/kotlin-doc.html#generate-the-documentation

**Comment by BraisGabin** [comment]
Should we link this documentation in the rule?

**Comment by schalkms** [comment]
> Should we link this documentation in the rule?

Yes, that's a good idea. I'll take care.

**Comment by arhont375** [comment]
Thank you for quick reply, but I'm afraid we misunderstood each other.
My issue that inline property documentation get reported with `UndocumentedPublicProperty`, I'm not sure why you suggesting me to take a look at the `@constructor` tag.

Unfortunately Dokka lacking proper examples, so I just rendered documentation to see if inline property documentation recognized or not and it had no issued:
For slightly changed example (so it's easier to understand from where docs are coming):
```kotlin
/**
 * This class have some KDoc.
 *
 * @property anotherImportantProperty Another important property documentation
 */
class TestProperties(
    /**
     * Serious property inline documentation
     */
    val seriousProperty: String,
    val anotherImportantProperty: Int,
) {
    /**
     * Some documentation
     */
    val justProperty: String = seriousProperty + seriousProperty

    fun importantUsage(): String = "$seriousProperty".repeat(anotherImportantProperty) + justProperty
}
```

I got following result:

![image](https://user-images.githubusercontent.com/6061346/114647876-c3854880-9d18-11eb-8ed9-1e0586b86e0f.png)

May you please reconsider this issue?

**Comment by schalkms** [comment]
@BraisGabin what do you think? I think the mentioned snippet is really rare and not mentioned in the KDoc. I'm not sure whether it's worth to change the rule in order to support this edge case.

**Comment by BraisGabin** [comment]
It is a false positive so I think the best approach is to reopen it and add the help wanted tag.

**Comment by cortinico** [comment]
> @BraisGabin what do you think? I think the mentioned snippet is really rare and not mentioned in the KDoc.

I believe the rule is behaving correctly, or at least should keep on reporting those cases. 
The KDoc guide (as described here https://kotlinlang.org/docs/kotlin-doc.html#property-name) mentions that `@property` should be used to describe constructor properties _where putting a doc comment directly before the property definition would be awkward._

I would still prefer Detekt to warn such comments to invite users to use `@property` correctly.

I think we should potentially add a more descriptive error message in this scenario but still raise an issue in some form (potentially we could also create a new rule for just this if it falls outside the scope of `UndocumentedPublicProperty`).

**Comment by BraisGabin** [comment]
> This tag can be used for documenting properties declared in the primary constructor, where putting a doc comment directly before the property definition would be awkward.

My english is far from perfect but what I understand here is that the use of `@property` is optional. The documentation doesn't encorage any of them. For me it sounds like: "If it looks awkward in the property, you can use `@property` instead"

For me, if someone wants to invest time in this issue I would happily merge it but I'll not expend time on it.

**Comment by cortinico** [comment]
> For me, if someone wants to invest time in this issue I would happily merge it but I'll not expend time on it.

Agree 👍 
As you said, the guide is not suggesting any of the two. 

My personal point of view on the issue is that I'd suggest the usage of `@property` instead. Mostly because I think the inline KDoc looks awkward most of the cases.

**Comment by arhont375** [comment]
> My personal point of view on the issue is that I'd suggest the usage of @property instead. Mostly because I think the inline KDoc looks awkward most of the cases.

Maybe best option would be to provide switch for this rule then, to make users to decide what's appropriate for them.

**Comment by arhont375** [comment]
Submitted PR to resolve this, not sure who should be assigned though (maybe you can set codeowners?).
Please take a look when you have time
